File: python/text_generation.py
# -*- coding: utf-8 -*-
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import pathlib
from absl.testing import absltest

media = pathlib.Path(__file__).parents[1] / "third_party"


class UnitTests(absltest.TestCase):

    def test_text_gen_text_only_prompt(self):
        # [START text_gen_text_only_prompt]
        from google import genai

        client = genai.Client()
        response = client.models.generate_content(
            model="gemini-3.8-flash", contents="Write a story about a magic backpack."
        )
        print(response.text)
        # [END text_gen_text_only_prompt]

    def test_text_gen_text_only_prompt_streaming(self):
        # [START text_gen_text_only_prompt_streaming]
        from google import genai

        client = genai.Client()
        response = client.models.generate_content_stream(
            model="gemini-3.8-flash", contents="Write a story about a magic backpack."
        )
        for chunk in response:
            print(chunk.text)
            print("_" * 80)
        # [END text_gen_text_only_prompt_streaming]

    def test_text_gen_multimodal_one_image_prompt(self):
        # [START text_gen_multimodal_one_image_prompt]
        from google import genai
        import PIL.Image

        client = genai.Client()
        organ = PIL.Image.open(media / "organ.jpg")
        response = client.models.generate_content(
            model="gemini-3.8-flash", contents=["Tell me about this instrument", organ]
        )
        print(response.text)
        # [END text_gen_multimodal_one_image_prompt]

    def test_text_gen_multimodal_one_image_prompt_streaming(self):
        # [START text_gen_multimodal_one_image_prompt_streaming]
        from google import genai
        import PIL.Image

        client = genai.Client()
        organ = PIL.Image.open(media / "organ.jpg")
        response = client.models.generate_content_stream(
            model="gemini-3.8-flash", contents=["Tell me about this instrument", organ]
        )
        for chunk in response:
            print(chunk.text)
            print("_" * 80)
        # [END text_gen_multimodal_one_image_prompt_streaming]

    def test_text_gen_multimodal_multi_image_prompt(self):
        # [START text_gen_multimodal_multi_image_prompt]
        from google import genai
        import PIL.Image

        client = genai.Client()
        organ = PIL.Image.open(media / "organ.jpg")
        cajun_instrument = PIL.Image.open(media / "Cajun_instruments.jpg")
        response = client.models.generate_content(
            model="gemini-3.8-flash",
            contents=[
                "What is the difference between both of these instruments?",
                organ,
                cajun_instrument,
            ],
        )
        print(response.text)
        # [END text_gen_multimodal_multi_image_prompt]

    def test_text_gen_multimodal_multi_image_prompt_streaming(self):
        # [START text_gen_multimodal_multi_image_prompt_streaming]
        from google import genai
        import PIL.Image

        client = genai.Client()
        organ = PIL.Image.open(media / "organ.jpg")
        cajun_instrument = PIL.Image.open(media / "Cajun_instruments.jpg")
        response = client.models.generate_content_stream(
            model="gemini-3.8-flash",
            contents=[
                "What is the difference between both of these instruments?",
                organ,
                cajun_instrument,
            ],
        )
        for chunk in response:
            print(chunk.text)
            print("_" * 80)
        # [END text_gen_multimodal_multi_image_prompt_streaming]

    def test_text_gen_multimodal_audio(self):
        # [START text_gen_multimodal_audio]
        from google import genai

        client = genai.Client()
        sample_audio = client.files.upload(file=media / "sample.mp3")
        response = client.models.generate_content(
            model="gemini-3.8-flash",
            contents=["Give me a summary of this audio file.", sample_audio],
        )
        print(response.text)
        # [END text_gen_multimodal_audio]

    def test_text_gen_multimodal_audio_streaming(self):
        # [START text_gen_multimodal_audio_streaming]
        from google import genai

        client = genai.Client()
        sample_audio = client.files.upload(file=media / "sample.mp3")
        response = client.models.generate_content_stream(
            model="gemini-3.8-flash",
            contents=["Give me a summary of this audio file.", sample_audio],
        )
        for chunk in response:
            print(chunk.text)
            print("_" * 80)
        # [END text_gen_multimodal_audio_streaming]

    def test_text_gen_multimodal_video_prompt(self):
        # [START text_gen_multimodal_video_prompt]
        from google import genai
        import time

        client = genai.Client()
        # Video clip (CC BY 3.0) from https://peach.blender.org/download/
        myfile = client.files.upload(file=media / "Big_Buck_Bunny.mp4")
        print(f"{myfile=}")

        # Poll until the video file is completely processed (state becomes ACTIVE).
        while not myfile.state or myfile.state.name != "ACTIVE":
            print("Processing video...")
            print("File state:", myfile.state)
            time.sleep(5)
            myfile = client.files.get(name=myfile.name)

        response = client.models.generate_content(
            model="gemini-3.8-flash", contents=[myfile, "Describe this video clip"]
        )
        print(f"{response.text=}")
        # [END text_gen_multimodal_video_prompt]

    def test_text_gen_multimodal_video_prompt_streaming(self):
        # [START text_gen_multimodal_video_prompt_streaming]
        from google import genai
        import time

        client = genai.Client()
        # Video clip (CC BY 3.0) from https://peach.blender.org/download/
        myfile = client.files.upload(file=media / "Big_Buck_Bunny.mp4")
        print(f"{myfile=}")

        # Poll until the video file is completely processed (state becomes ACTIVE).
        while not myfile.state or myfile.state.name != "ACTIVE":
            print("Processing video...")
            print("File state:", myfile.state)
            time.sleep(5)
            myfile = client.files.get(name=myfile.name)

        response = client.models.generate_content_stream(
            model="gemini-3.8-flash", contents=[myfile, "Describe this video clip"]
        )
        for chunk in response:
            print(chunk.text)
            print("_" * 80)
        # [END text_gen_multimodal_video_prompt_streaming]

    def test_text_gen_multimodal_pdf(self):
        # [START text_gen_multimodal_pdf]
        from google import genai

        client = genai.Client()
        sample_pdf = client.files.upload(file=media / "test.pdf")
        response = client.models.generate_content(
            model="gemini-3.8-flash",
            contents=["Give me a summary of this document:", sample_pdf],
        )
        print(f"{response.text=}")
        # [END text_gen_multimodal_pdf]

    def test_text_gen_multimodal_pdf_streaming(self):
        # [START text_gen_multimodal_pdf_streaming]
        from google import genai

        client = genai.Client()
        sample_pdf = client.files.upload(file=media / "test.pdf")
        response = client.models.generate_content_stream(
            model="gemini-3.8-flash",
            contents=["Give me a summary of this document:", sample_pdf],
        )

        for chunk in response:
            print(chunk.text)
            print("_" * 80)
        # [END text_gen_multimodal_pdf_streaming]


if __name__ == "__main__":
    absltest.main()

File: javascript/text_generation.js
/**
 * @license
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { 
  GoogleGenAI, 
  createUserContent, 
  createPartFromUri 
} from "@google/genai";
import path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const media = path.join(__dirname, '..', 'third_party');

// Sleep helper for video polling.
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

export async function textGenTextOnlyPrompt() {
  // [START text_gen_text_only_prompt]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

  const response = await ai.models.generateContent({
    model: "gemini-3.8-flash",
    contents: "Write a story about a magic backpack.",
  });
  console.log(response.text);
  // [END text_gen_text_only_prompt]
  return response.text;
}

export async function textGenTextOnlyPromptStreaming() {
  // [START text_gen_text_only_prompt_streaming]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

  const response = await ai.models.generateContentStream({
    model: "gemini-3.8-flash",
    contents: "Write a story about a magic backpack.",
  });
  let text = "";
  for await (const chunk of response) {
    console.log(chunk.text);
    text += chunk.text;
  }
  // [END text_gen_text_only_prompt_streaming]
  return text;
}

export async function textGenMultimodalOneImagePrompt() {
  // [START text_gen_multimodal_one_image_prompt]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

  const organ = await ai.files.upload({
    file: path.join(media, "organ.jpg"),
  });

  const response = await ai.models.generateContent({
    model: "gemini-3.8-flash",
    contents: [
      createUserContent([
        "Tell me about this instrument", 
        createPartFromUri(organ.uri, organ.mimeType)
      ]),
    ],
  });
  console.log(response.text);
  // [END text_gen_multimodal_one_image_prompt]
  return response.text;
}

export async function textGenMultimodalOneImagePromptStreaming() {
  // [START text_gen_multimodal_one_image_prompt_streaming]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

  const organ = await ai.files.upload({
    file: path.join(media, "organ.jpg"),
  });

  const response = await ai.models.generateContentStream({
    model: "gemini-3.8-flash",
    contents: [
      createUserContent([
        "Tell me about this instrument", 
        createPartFromUri(organ.uri, organ.mimeType)
      ]),
    ],
  });
  let text = "";
  for await (const chunk of response) {
    console.log(chunk.text);
    text += chunk.text;
  }
  // [END text_gen_multimodal_one_image_prompt_streaming]
  return text;
}

export async function textGenMultimodalMultiImagePrompt() {
  // [START text_gen_multimodal_multi_image_prompt]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });
  
  const organ = await ai.files.upload({
    file: path.join(media, "organ.jpg"),
  });

  const cajun = await ai.files.upload({
    file: path.join(media, "Cajun_instruments.jpg"),
    config: { mimeType: "image/jpeg" },
  });

  const response = await ai.models.generateContent({
    model: "gemini-3.8-flash",
    contents: [
      createUserContent([
        "What is the difference between both of these instruments?",
        createPartFromUri(organ.uri, organ.mimeType),
        createPartFromUri(cajun.uri, cajun.mimeType),
      ]),
    ],
  });
  console.log(response.text);
  // [END text_gen_multimodal_multi_image_prompt]
  return response.text;
}

export async function textGenMultimodalMultiImagePromptStreaming() {
  // [START text_gen_multimodal_multi_image_prompt_streaming]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

  const organ = await ai.files.upload({
    file: path.join(media, "organ.jpg"),
  });

  const cajun = await ai.files.upload({
    file: path.join(media, "Cajun_instruments.jpg"),
  });

  const response = await ai.models.generateContentStream({
    model: "gemini-3.8-flash",
    contents: [
      createUserContent([
        "What is the difference between both of these instruments?",
        createPartFromUri(organ.uri, organ.mimeType),
        createPartFromUri(cajun.uri, cajun.mimeType),
      ]),
    ],
  });
  let text = "";
  for await (const chunk of response) {
    console.log(chunk.text);
    text += chunk.text;
  }
  // [END text_gen_multimodal_multi_image_prompt_streaming]
  return text;
}

export async function textGenMultimodalAudio() {
  // [START text_gen_multimodal_audio]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

  const audio = await ai.files.upload({
    file: path.join(media, "sample.mp3"),
  });
  
  const response = await ai.models.generateContent({
    model: "gemini-3.8-flash",
    contents: [
      createUserContent([
        "Give me a summary of this audio file.",
        createPartFromUri(audio.uri, audio.mimeType),
      ]),
    ],
  });
  console.log(response.text);
  // [END text_gen_multimodal_audio]
  return response.text;
}

export async function textGenMultimodalAudioStreaming() {
  // [START text_gen_multimodal_audio_streaming]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

  const audio = await ai.files.upload({
    file: path.join(media, "sample.mp3"),
  });

  const response = await ai.models.generateContentStream({
    model: "gemini-3.8-flash",
    contents: [
      createUserContent([
        "Give me a summary of this audio file.",
        createPartFromUri(audio.uri, audio.mimeType),
      ]),
    ],
  });
  let text = "";
  for await (const chunk of response) {
    console.log(chunk.text);
    text += chunk.text;
  }
  // [END text_gen_multimodal_audio_streaming]
  return text;
}

export async function textGenMultimodalVideoPrompt() {
  // [START text_gen_multimodal_video_prompt]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

  let video = await ai.files.upload({
    file: path.join(media, 'Big_Buck_Bunny.mp4'),
  });

  // Poll until the video file is completely processed (state becomes ACTIVE).
  while (!video.state || video.state.toString() !== 'ACTIVE') {
    console.log('Processing video...');
    console.log('File state: ', video.state);
    await sleep(5000);
    video = await ai.files.get({name: video.name});
  }

  const response = await ai.models.generateContent({
    model: "gemini-3.8-flash",
    contents: [
      createUserContent([
        "Describe this video clip",
        createPartFromUri(video.uri, video.mimeType),
      ]),
    ],
  });
  console.log(response.text);
  // [END text_gen_multimodal_video_prompt]
  return response.text;
}

export async function textGenMultimodalVideoPromptStreaming() {
  // [START text_gen_multimodal_video_prompt_streaming]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

  let video = await ai.files.upload({
    file: path.join(media, 'Big_Buck_Bunny.mp4'),
  });

  // Poll until the video file is completely processed (state becomes ACTIVE).
  while (!video.state || video.state.toString() !== 'ACTIVE') {
    console.log('Processing video...');
    console.log('File state: ', video.state);
    await sleep(5000);
    video = await ai.files.get({name: video.name});
  }

  const response = await ai.models.generateContentStream({
    model: "gemini-3.8-flash",
    contents: [
      createUserContent([
        "Describe this video clip",
        createPartFromUri(video.uri, video.mimeType),
      ]),
    ],
  });
  let text = "";
  for await (const chunk of response) {
    console.log(chunk.text);
    text += chunk.text;
  }
  // [END text_gen_multimodal_video_prompt_streaming]
  return text;
}

export async function textGenMultimodalPdf() {
  // [START text_gen_multimodal_pdf]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

  const pdf = await ai.files.upload({
    file: path.join(media, "test.pdf"),
  });

  const response = await ai.models.generateContent({
    model: "gemini-3.8-flash",
    contents: [
      createUserContent([
        "Give me a summary of this document:",
        createPartFromUri(pdf.uri, pdf.mimeType),
      ]),
    ],
  });
  console.log(response.text);
  // [END text_gen_multimodal_pdf]
  return response.text;
}

export async function textGenMultimodalPdfStreaming() {
  // [START text_gen_multimodal_pdf_streaming]
  // Make sure to include the following import:
  // import {GoogleGenAI} from '@google/genai';
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

  const pdf = await ai.files.upload({
    file: path.join(media, "test.pdf"),
  });

  const response = await ai.models.generateContentStream({
    model: "gemini-3.8-flash",
    contents: [
      createUserContent([
        "Give me a summary of this document:",
        createPartFromUri(pdf.uri, pdf.mimeType),
      ]),
    ],
  });
  let text = "";
  for await (const chunk of response) {
    console.log(chunk.text);
    text += chunk.text;
  }
  // [END text_gen_multimodal_pdf_streaming]
  return text;
}
